libc/signal/handler.c

```diff
@@ -0,0 +1,302 @@
+
+
+	TEST_ASSERT_EQUAL_INT(0, sigprocmask(SIG_SETMASK, NULL, &set));
+
+	/* after the signal handler - the original mask should be restored - only SIGUSR1 blocked */
+	TEST_ASSERT_EQUAL_INT(0, sigismember(&set, SIGPIPE));
+	TEST_ASSERT_EQUAL_INT(1, sigismember(&set, SIGUSR1));
+	TEST_ASSERT_EQUAL_INT(0, sigismember(&set, SIGUSR2));
+}
+
+
+/* check mask consistiency on signal handler - whether signal handler will be called if the signal is only unblocked in sa_mask of the other signal
+ * should not be called as sa_mask is ORed with current thread signal mask */
+TEST(handler, sighandler_signal_in_signal)
+{
+	sigset_t set, oldset;
+	struct sigaction sa = { 0 };
+
+	/* SIGUSR2 - block in normal execution, unblock in sa_mask */
+	TEST_ASSERT_EQUAL_INT(0, sigemptyset(&set));
+	TEST_ASSERT_EQUAL_INT(0, sigaddset(&set, SIGUSR2));
+	TEST_ASSERT_EQUAL_INT(0, sigprocmask(SIG_SETMASK, &set, NULL));
+
+	sa.sa_handler = sighandler;
+	sa.sa_flags = 0;
+
+	TEST_ASSERT_EQUAL_INT(0, sigfillset(&sa.sa_mask));
+	TEST_ASSERT_EQUAL_INT(0, sigdelset(&sa.sa_mask, SIGUSR2));
+
+	handler_haveSignal = 0u;
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGPIPE, &sa, NULL));
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGUSR2, &sa, NULL));
+
+	/* send SIGUSR2, verify it's pending */
+	TEST_ASSERT_EQUAL_INT(0, kill(getpid(), SIGUSR2));
+	TEST_ASSERT_EQUAL_HEX32(0, handler_haveSignal);
+
+	TEST_ASSERT_EQUAL_INT(0, kill(getpid(), SIGPIPE));
+
+	/* signal handler should be called on syscall exit from kill - only SIGPIPE should be delivered */
+	TEST_ASSERT_EQUAL_INT((1u << SIGPIPE), handler_haveSignal);
+
+
+	/* unblock SIGUSR2 */
+	handler_haveSignal = 0u;
+	TEST_ASSERT_EQUAL_INT(0, sigdelset(&set, SIGUSR2));
+	TEST_ASSERT_EQUAL_INT(0, sigprocmask(SIG_SETMASK, &set, &oldset));
+
+	/* SIGUSR2 should be delivered now */
+	TEST_ASSERT_EQUAL_HEX32((1u << SIGUSR2), handler_haveSignal);
+
+	/* after the signal handler - the original mask should be restored - only SIGUSR2 blocked */
+	TEST_ASSERT_EQUAL_INT(0, sigismember(&oldset, SIGPIPE));
+	TEST_ASSERT_EQUAL_INT(0, sigismember(&oldset, SIGUSR1));
+	TEST_ASSERT_EQUAL_INT(1, sigismember(&oldset, SIGUSR2));
+}
+
+/* TODO: test sa_flags - especially SA_NODEFER */
+
+
+TEST_GROUP_RUNNER(handler)
+{
+	RUN_TEST_CASE(handler, sighandler_sa_mask);
+	RUN_TEST_CASE(handler, sighandler_signal_in_signal);
+}
+
+
+TEST_GROUP(sigsuspend);
+
+
+TEST_SETUP(sigsuspend)
+{
+	struct sigaction sa = {
+		.sa_handler = sighandler,
+	};
+
+	sigemptyset(&sa.sa_mask);
+	TEST_ASSERT_EQUAL_INT(0, sigaddset(&sa.sa_mask, SIGPIPE));
+
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGALRM, &sa, NULL));
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGUSR1, &sa, NULL));
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGUSR2, &sa, NULL));
+}
+
+
+TEST_TEAR_DOWN(sigsuspend)
+{
+	/* unblock all signals */
+	sigset_t set;
+	sigemptyset(&set);
+	sigprocmask(SIG_SETMASK, &set, NULL);
+
+	/* set default signal disposition for all signals */
```

Comment:
Maybe comment should be adapted, 3 dispositions are changed not all

Reply:
mental shortcut, `all signals` mean all signals for which signal handler was installed a few lines above. Added longer description to be unambiguous.